src/context.ml

```diff
@@ -599,32 +597,73 @@ let create_for_opam ~root ~env ~env_nodes ~targets ~profile
   in
   let env = Env.extend env ~vars in
   create ~kind:(Opam { root; switch }) ~profile ~targets ~path ~env ~env_nodes
-    ~name ~merlin ~host_toolchain
-
-let create ~env (workspace : Workspace.t) =
-  let env_nodes context =
+    ~name ~merlin ~host_context ~host_toolchain
+
+let separate_independant_contexts topo_contexts =
+  let f (acc, cur) elem =
+    match Workspace.Context.host_context elem, cur with
+    | None, [] -> (acc, [elem])
+    | None, _ -> ((List.rev cur) :: acc, [elem])
+    | Some _, _ -> (acc, elem :: cur)
```

Comment:
I'm finding it hard to understand the logic here. Another approach would be to add the following function to `Fiber`:

```ocaml
module type Memory = sig
  type key
  type 'a t
  val empty : 'a t
  val add : 'a t -> key -> 'a -> 'a t
  val find : 'a t -> key -> 'a option
end

(** [memoize memory f] returns a version of [f] that remembers the
    result of previous call to [f] in order to not recompute them. *)
val memoize
  :  (module Memory with type key := 'key)
  -> ('a -> 'b t)
  -> ('a -> 'b t) Staged.t
```

and then memoize the instantiation function before passing it to `parallel_map`.

Reply:
I've been thinking about this as well, but I'm wondering why do we need anything extra here. In the previous version of this feature, we managed without any topological sorting by doing something like this:

```ocaml
let rec contexts = lazy (
  List.map workspace.contexts ~f:(fun context ->
    let context =
      let* host = String.Map.find_exn (Lazy.force contexts) context.host in
      create_context ~context ~host
    in
    (context.name, context))
  |> String.Map.of_list_exn
) in
let* contexts = Lazy.force contexts |> String.Map.values |> Fiber.all in
...
```

Actually, I'm pretty sure we used a hashtable before, but I don't see why a map can't be used.